src/database/pvRecord.cpp

```diff
@@ -77,22 +77,28 @@ void PVRecord::destroy()
         isDestroyed = true;
         pvTimeStamp.detach();
     
-        std::list<PVRecordClientPtr>::iterator clientIter;
+        std::list<PVRecordClientWPtr>::iterator clientIter;
         while(true) {
             clientIter = pvRecordClientList.begin();
             if(clientIter==pvRecordClientList.end()) break;
+            PVRecordClientPtr client = clientIter->lock();
+            if(client.get()) {
+                unlock();
+                client->detach(getPtrSelf());
+                lock();
+            }
             pvRecordClientList.erase(clientIter);
```

Comment:
On 01/19/2016 11:47 AM, mdavidsaver wrote:
>
> In src/database/pvRecord.cpp 
> <https://github.com/epics-base/pvDatabaseCPP/pull/4#discussion_r50138258>:
>
> >          while(true) {
> >              clientIter = pvRecordClientList.begin();
> >              if(clientIter==pvRecordClientList.end()) break;
> > +            PVRecordClientPtr client = clientIter->lock();
> > +            if(client.get()) {
> > +                unlock();
> > +                client->detach(getPtrSelf());
> > +                lock();
> > +            }
> >              pvRecordClientList.erase(clientIter);

This appears in

> void PVRecord::destroy()
> {
>     if(traceLevel>1) {
>         cout << "PVRecord::destroy() " << recordName << endl;
>     }
>     lock();
>     try {
>         if(isDestroyed) {
>             unlock();
>             return;
>         }
>         isDestroyed = true;
>         pvTimeStamp.detach();
>
>         std::list<PVRecordClientWPtr>::iterator clientIter;
>         while(true) {
>             clientIter = pvRecordClientList.begin();
>             if(clientIter==pvRecordClientList.end()) break;
>             PVRecordClientPtr client = clientIter->lock();
>             if(client.get()) {
>                 unlock();
>                 client->detach(getPtrSelf());
>                 lock();
>             }
>             pvRecordClientList.erase(clientIter);
>         }
>         std::list<PVListenerWPtr>::iterator listenerIter;
>         while(true) {
>             listenerIter = pvListenerList.begin();
>             if(listenerIter==pvListenerList.end()) break;
>             PVListenerPtr listener = listenerIter->lock();
>             if(listener.get()) {
>                 unlock();
>                 listener->unlisten(getPtrSelf());
>                 lock();
>             }
>             pvListenerList.erase(listenerIter);
>
>         }
>         pvRecordStructure->destroy();
>         pvRecordStructure.reset();
>         pvStructure.reset();
>         unlock();
>     } catch(...) {
>         unlock();
>         throw;
>     }
> }

So the code IS enclosed in
try {
    ...
} catch(...) {
         unlock();
         throw;
     }

Note also that  the way lock is implemented it is OK to call unlock 
twice after a call to lock.
See pvDataCPP/src/misc/lock.h





> The most obvious issue here is that explicit unlock()/lock() leaves 
> open the possibility that an exception thrown be 'client->detach()' 
> would no re-lock. There are several ways to correct this. One is 
> explicit try/catch/rethrow.
>
> if(client) {
>    unlock();
>    try  {
>      client->detach(getPtrSelf());
>      lock();
>    }catch(...) {
>      lock();
>      throw;
>    }
> }
>
> I personally find this verbose and potentially error prone as the 
> re-lock() call should be duplicated. My preference is to add a "guard" 
> class to handle unlock() and re-lock(). So a definition is added to 
> some header.
>
> template<typename  T>
> struct  XUnLocker  {
>    T* ref;
>    XUnLocker(T& r) :ref(&r) { r.unlock(); }
>   ~XUnLocker() { r.lock(); }
> };
>
> Then the code in question becomes:
>
> if(client) {
>    XUnLocker<PVRecord>L(*this);// calls unlock()
>    client->detach(getPtrSelf());
>    // lock() when 'L' goes out of scope due to normal flow or exception 
> unwind
> }
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/epics-base/pvDatabaseCPP/pull/4/files#r50138258>.
>